src/DotNetProjectFile.Analyzers/Analyzers/MsBuild/ThirdPartyLicenseResolver.cs

```diff
@@ -7,32 +6,59 @@ namespace DotNetProjectFile.Analyzers.MsBuild;
 [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
 public sealed class ThirdPartyLicenseResolver() : MsBuildProjectFileAnalyzer(
     Rule.OnlyIncludePackagesWithExplicitLicense,
-    Rule.PackageOnlyContainsDeprecatedLicenseUrl)
+    Rule.PackageOnlyContainsDeprecatedLicenseUrl,
+    Rule.PackageContainsIncompatibleLicense)
 {
     /// <inheritdoc />
     public override bool DisableOnFailingImport => false;
 
     /// <inheritdoc />
-    protected override void Register(ProjectFileAnalysisContext<MsBuildProject> context)
+    protected override void Register(ProjectFileAnalysisContext context)
     {
+        var allowed = (context.GetMsBuildProperty("AllowedLicenses") ?? string.Empty)
+            .Split([',', ';'], StringSplitOptions.RemoveEmptyEntries)
+            .Select(l => Licenses.FromExpression(l.Trim()))
+            .ToArray();
+
         foreach (var reference in context.File.ItemGroups.SelectMany(g => g.Children)
             .OfType<PackageReferenceBase>()
             .Where(r => r.Version is { Length: > 0 }))
         {
-            if (reference.GetLicensedPackage() is not { } package)
-            {
-                context.ReportDiagnostic(Rule.OnlyIncludePackagesWithExplicitLicense, reference, reference.IncludeOrUpdate);
-            }
-            else if (package.UrlOnly() && Licenses.FromUrl(package.LicenseUrl) == Licenses.Unknown)
-            {
-                context.ReportDiagnostic(Rule.PackageOnlyContainsDeprecatedLicenseUrl, reference, reference.IncludeOrUpdate);
-            }
+            Report(reference, allowed, context);
+        }
+    }
+
+    private static void Report(PackageReferenceBase reference, LicenseExpression[] allowed, ProjectFileAnalysisContext context)
+    {
+        if (reference.GetLicensedPackage() is not { } package)
+        {
+            context.ReportDiagnostic(Rule.OnlyIncludePackagesWithExplicitLicense, reference, reference.IncludeOrUpdate);
+            return;
+        }
+
+        var expression = package.LicenseExpression();
+
+        if (package.UrlOnly() && expression.IsUnknown)
+        {
+            context.ReportDiagnostic(Rule.PackageOnlyContainsDeprecatedLicenseUrl, reference, reference.IncludeOrUpdate);
+        }
+        else if (expression.IsKnown && allowed.None(l => l.CompatibleWith(expression)))
```

Comment:
This is checking that at least one allowed license could be used a dependency for the dependency that is being analyzed. I don't think that is what you want to check here.

I expect you would want to either check that `!allowed.Contains(expression) || !expression.CompatibleWith(<ParsedLicense-node-of-this-project>)`

Reply:
`a.CompatibleWith(b)` checks that a project with license `b` is allowed to have a package with license `a` as a dependency.

Reply:
For example:

```csharp
GPL3.CompatibleWith(AGPL3) // true
AGPL3.CompatibleWith(GPL3) // false

GPL3.CompatibleWith(MIT) // false
MIT.CompatibleWith(GPL3) // true

BSD3.CompatibleWith(MIT) // true
MIT.CompatibleWith(BSD3) // true

GPL3.CompatibleWith(GPL2) // false
GPL2.CompatibleWith(GPL3) // false
```

Reply:
The second part of the expression is a really nice idea! It should have a (slightly) different message I think. It should be explicit that the reference is incompatible with the license of the project's package it self, rather than being not being allowed because it is not in some fixed list.